prov/efa/test/efa_unit_test_rdm_peer.c

```diff
@@ -0,0 +1,266 @@
+	}
+
+	(&(*peer)->robuf)->exp_msg_id = efa_env.recvwin_size;
+	efa_env.rx_copy_ooo = 0;
+	efa_rdm_peer_move_overflow_pke_to_recvwin(*peer);
+}
+
+void test_efa_rdm_peer_move_overflow_pke_to_recvwin(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	struct efa_rdm_peer *peer;
+	struct efa_rdm_pke *pkt_entry;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	/* overflow_pke_list has a pkt entry with msg_id 18000.
+	 * After calling efa_rdm_peer_move_overflow_pke_to_recvwin when exp_msg_id = 16384,
+	 * 18000 will be moved to recvwin and overflow_pke_list will be empty. */
+	test_efa_rdm_peer_move_overflow_pke_to_recvwin_impl(
+		resource, 18000, &peer, &pkt_entry, 1,
+		EFA_RDM_EAGER_MSGRTM_PKT);
+
+	assert_non_null(*ofi_recvwin_get_msg((&peer->robuf), 18000));
+	assert_int_equal(efa_unit_test_get_dlist_length(&peer->overflow_pke_list), 0);
+
+	efa_rdm_pke_release_rx(pkt_entry);
+}
+
+void test_efa_rdm_peer_keep_pke_in_overflow_list(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	struct efa_rdm_peer *peer;
+	struct efa_rdm_pke *pkt_entry;
+	struct efa_rdm_peer_overflow_pke_list_entry *overflow_pke_list_entry;
+	struct dlist_entry *tmp;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	/* overflow_pke_list has a pkt entry with msg_id 33000.
+	 * After calling efa_rdm_peer_move_overflow_pke_to_recvwin when exp_msg_id = 16384,
+	 * 33000 will stay in overflow_pke_list. */
+	test_efa_rdm_peer_move_overflow_pke_to_recvwin_impl(
+		resource, 33000, &peer, &pkt_entry, 1,
+		EFA_RDM_EAGER_MSGRTM_PKT);
+
+	assert_int_equal(efa_unit_test_get_dlist_length(&peer->overflow_pke_list), 1);
+
+	dlist_foreach_container_safe (
+		&peer->overflow_pke_list,
+		struct efa_rdm_peer_overflow_pke_list_entry,
+		overflow_pke_list_entry, entry, tmp) {
+		dlist_remove(&overflow_pke_list_entry->entry);
+		efa_rdm_pke_release_rx(overflow_pke_list_entry->pkt_entry);
+		ofi_buf_free(overflow_pke_list_entry);
+	}
+}
+
+void test_efa_rdm_peer_append_overflow_pke_to_recvwin(struct efa_resource **state) {
+	struct efa_resource *resource = *state;
+	struct efa_rdm_peer *peer;
+	struct efa_rdm_pke *pkt_entry;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	/* overflow_pke_list has two pkt entries with msg_id 17000.
+	* After calling efa_rdm_peer_move_overflow_pke_to_recvwin when exp_msg_id = 16384,
+	* two pkt entries of same msg id will be appended to the same entry in recvwin,
+	* and overflow_pke_list will be empty. */
+	test_efa_rdm_peer_move_overflow_pke_to_recvwin_impl(
+		resource, 17000, &peer, &pkt_entry, 2,
+		EFA_RDM_MEDIUM_TAGRTM_PKT);
+
+	pkt_entry = *ofi_recvwin_get_msg((&peer->robuf), 17000);
+	assert_non_null(pkt_entry);
+	assert_non_null(pkt_entry->next);
+	assert_int_equal(efa_unit_test_get_dlist_length(&peer->overflow_pke_list), 0);
+
+	pkt_entry->next = NULL;
+	efa_rdm_pke_release_rx(pkt_entry);
```

Comment:
Will both pkes be released?

Reply:
I am not sure what is the right way to release both.
I also tried 
```
pkt_entry->next->next = NULL;
efa_rdm_pke_release_rx(pkt_entry->next);
pkt_entry->next = NULL;
efa_rdm_pke_release_rx(pkt_entry);
```
but got ofi_bufpool_destroy: Assertion `(pool->attr.flags & OFI_BUFPOOL_NO_TRACK) || !ofi_atomic_get32(&buf_region->use_cnt)` failed.

Reply:
You call efa_rdm_pke_alloc twice while assigning them to the same `*pkt_entry`, shouldn't you use a vector here?